File: src/frontends/pepc-breakup/frand123-master/examples/Fortran/integer32.f90
!
! Example illustrating the use of the frand123 Fortran interface for the
! generation of 32-bit signed integer random numbers.
! The signed integer random numbers ar used as experiment for coin flip.
!

program integer32
   use, intrinsic :: iso_c_binding, only: c_int64_t, c_int32_t
   use frand123, only: frand123_state_kind, frand123_state_size, &
                       frand123Init, frand123Integer32
   implicit none

   ! state for frand123
   integer(kind=frand123_state_kind), dimension(frand123_state_size) :: state

   ! init state
   call frand123Init( state, 0, 0 )

   ! use coinFlipScalar
   write(*,*) 'Generate one random number at a time'
   call coinFlipScalar( state )

   ! use coinFlip10
   write(*,*) 'Generate all 10 random numbers afront'
   call coinFlip10( state )

contains

   ! simulate coin flip by generating one random number at a time
   subroutine coinFlipScalar( state )
      implicit none
      integer( kind = frand123_state_kind ), dimension( frand123_state_size ), &
                                             intent( inout ) :: state

      ! result of coin flip
      integer( kind = c_int32_t ) :: coin

      ! iteration variable
      integer :: i

      ! simulate 10 coin flips
      ! > 0 heads
      ! < 0 tails
      ! = 0 what a luck
      do i = 1, 10
         ! create 32-bit signed random integer
         call frand123Integer32( state, coin )
         ! result
         if( coin .eq. 0 ) then
            write(*,*) 'What a luck'
         elseif( coin .gt. 0 ) then
            write(*,*) 'heads'
         else
            write(*,*) 'tails'
         endif
      enddo
   end subroutine coinFlipScalar

   ! simulate coin flip by generating all random numbers afront
   subroutine coinFlip10( state )
      implicit none
      integer( kind = frand123_state_kind ), dimension( frand123_state_size ), &
                                             intent( inout ) :: state

      ! result of coin flip
      integer( kind = c_int32_t ), dimension( 10 ) :: coin

      ! iteration variable
      integer :: i

      ! create 32-bit signed random integers
      call frand123Integer32( state, coin )

      ! simulate 10 coin flips
      ! > 0 heads
      ! < 0 tails
      ! = 0 what a luck
      do i = 1, 10
         ! result
         if( coin( i ) .eq. 0 ) then
            write(*,*) 'What a luck'
         elseif( coin( i ) .gt. 0 ) then
            write(*,*) 'heads'
         else
            write(*,*) 'tails'
         endif
      enddo
   end subroutine coinFlip10
end program integer32

File: src/frontends/pepc-breakup/frand123-master/examples/C/integer32.c
/*
 * Example illustrating the use of the frand123 C interface for the generation
 * of 32-bit signed and unsigned integer random numbers.
 * The unsigned integer random numbers are used as containers for ranom bits.
 * The signed integer random numbers are used as experiment for coin flip.
 */

#define __STDC_FORMAT_MACROS
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>
#include <stdbool.h>
#include "../../wrapper/frand123.h"

// check if the position-th bit is set in bitset
bool isBitSet( uint64_t bitset, short position )
{
   // create a mask by shifting 1 left by position, i.e. calculating 2^position
   uint32_t mask = UINT64_C( 1 ) << position;
   // compare to bitset
   uint32_t masked = bitset & mask;
   // return true if it is set
   return masked > 0;
}

// check whether certain bits are set in random unsigned 32-bit integers
void checkBitSets( int64_t *state )
{
   // generate five uint64_t as containers for random bits
   uint32_t randomBits[ 5 ];
   frand123UnsignedInteger32( state, 5, randomBits );
   
   // check each case
   for( int i = 1; i < 5; i++ )
   {
      printf( "isBitSet( %" PRIu32 ", 31): %s\n", randomBits[ i ], isBitSet( randomBits[ i ], 31 ) ? "true" : "false" );
      printf( "isBitSet( %" PRIu32 ",  0): %s\n", randomBits[ i ], isBitSet( randomBits[ i ], 0 ) ? "true" : "false" );
   }

   return;
}

// simulates a coin flip:
// > 0 heads
// < 0 tails
// = 0 what a luck
void naiveCoinFlip( int64_t *state )
{
   char heads[ 6 ] = "heads";
   char tails[ 6 ] = "tails";
   char zero[ 12 ] = "what a luck";
   for( int i = 0; i < 10; i++ )
   {
      int32_t flip = frand123Integer32_scalar( state );
      char *text;
      if( flip == 0 )
      {
         text = zero;
      }
      else if( flip > 0 )
      {
         text = heads;
      }
      else
      {
         text = tails;
      }
      printf( "Flip %d: %s\n", i, text );
   }
   return;
}

int main()
{
   // build state
   int64_t state[ 4 ];
   frand123Init( state, 0, 0, NULL );

   checkBitSets( state );

   printf( "\n\n" );

   naiveCoinFlip( state );

   return 0;
}
